src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -186,7 +186,14 @@ private Type0CommandMessageSection<BsonDocument> CreateType0Section(ConnectionDe
             }
             Action<BsonWriterSettings> writerSettingsConfigurator = s => s.GuidRepresentation = GuidRepresentation.Unspecified;
 
-            if (_session.IsInTransaction)
+            string commandName = null;
+            if (_command.GetType() == typeof(BsonDocument))
```

Comment:
This conditional seems risky, since the behavior will silently change based on the concrete type of _command.  
I chose a bit different path in the Java driver that I'm wondering if you'd like better.  There the session encapsulate the state of whether we're currently in the process of calling commitTransaction (there's a boolean that's set).  That way you don't need to pass the command name in since the session knows already.  Additionally, IsInTransaction returns true if the state is COMMITTED and we're in the process of calling commitTransaction.

Reply:
I meant to comment about this.

We *could* check the command name of *every* command, but if the command is something weird like a RawBsonDocument it would  be inefficient.

This is an optimization based on our knowledge that the abort and commit commands use normal BsonDocuments.

I'll think about the other part of your comment.

Reply:
All we need to know there is whether the command name is "commitTransaction".  Which we would know if the session were tracking whether a commitTransaction command is currently in progress.

Reply:
We also need to know if the command name is "abortTransaction", because we don't want to leave the "Aborted" or "Committed" states if the command name is *either* "commitTransaction" *or* "abortTransaction".

Or are you thinking that we would never get this far if the command was "abortTransaction"?

Reply:
Currently the AboutToSendCommand method only compares to "commitTransaction" not to "abortTransaction".  The current tests don't require any more than that.  The difference is that CommitTransaction can be retried, while AbortTransaction can not.

Reply:
True enough. Hence my question whether you were thinking (or agree with) that we would never get this far if the command was "abortTransaction". So it looks like in principle we could refactor AboutToSendCommand to not take a commandName.

But... we still need to know the commandName. Or somehow know whether we're retrying a commit. I could add a _commitInProgress field to CoreTransaction like you did, but then I would have to add something to the public API of ICoreSession to query it. For example:

    if (_session.IsInTransaction || _session.CurrentTransaction?.IsCommitInProgress)

I think I'd rather stick to peeking at the command name than add yet another property to the public API of ICoreSession.

Reply:
In Java the property is private.  But the hasActiveTransaction method/property takes it into account, so it's abstracted (thought admittedly it's weird either way)

Reply:
I've kept inspecting the command directly to see the command name as the simplest possible implementation of what's needed.

Reply:
I've spiked a version that does not look at the commandName, and uses your approach of having a _isCommitTransactionInProgress field in CoreSession.

The trick you suggested of having IsInTransaction take _isCommitTransactionInProgress into account is working.

In words this is like saying: once you call CommitTransaction you are no longer "in" a transaction, but if you retry the commit you are temporarily "back in" the already committed transaction for purposes of retrying the commit.

Do you want me to move to this approach and push a new commit?

Reply:
Yes, please, as I still think it's cleaner to encapsulate this logic in the session and clean up the code down in the wire protocol.